### PR TITLE
HF image generation that integrates with Cory's earlier script

### PR DESCRIPTION
**Pull request by rishab-partha**
Generates images at scale for an HF model from an HF dataset and patches those results to remote storage. Tested with 15 different HF models and with OCI.

### CODE REVIEW

File: diffusion/evaluation/generate_images.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                 guidance_scale: float = 7.0,
+                 height: int = 1024,
+                 width: int = 1024,
+                 caption_key: str = 'caption',
+                 seed: int = 17,
+                 output_bucket: Optional[str] = None,
+                 output_prefix: Optional[str] = None,
+                 extra_keys: Optional[list] = None,
+                 additional_generate_kwargs: Optional[Dict] = None,
+                 hf_model: Optional[bool] = False):
```
Comment (coryMosaicML) [inline diffusion/evaluation/generate_images.py:59]: Could you add this to the docstring?

File: diffusion/evaluation/generate_images.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        if isinstance(model, str) and hf_model == False:
+            raise ValueError('Can only use strings for model with hf models!')
+        self.hf_model = hf_model
+        if hf_model or isinstance(model, str):
+            print(f'LOCALRANK{dist.get_local_rank()}')
```
Comment (coryMosaicML) [inline diffusion/evaluation/generate_images.py:65]: Maybe a leftover print?

File: diffusion/evaluation/generate_images.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            if dist.get_local_rank() == 0:
```
Comment (coryMosaicML) [inline diffusion/evaluation/generate_images.py:66]: Is this to ensure checkpoint download only happens on local rank 0, and the subsequent `from_pretrained` call doesn't trigger another download?

File: diffusion/evaluation/generate_images.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    def generate(self):
+        """Core image generation function. Generates images at a given guidance scale.
+
+        Args:
+            guidance_scale (float): The guidance scale to use for image generation.
+        """
+        os.makedirs(os.path.join('/tmp', self.output_prefix), exist_ok=True)
+        # Partition the dataset across the ranks
+        if self.hf_model:
+            dataset_len = self.dataset.num_rows  # type: ignore
```
Comment (coryMosaicML) [inline diffusion/evaluation/generate_images.py:120]: Is this for using with a hf dataset?

File: diffusion/generate.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    if not config.hf_model:
+        model: ComposerModel = hydra.utils.instantiate(config.model)
+    else:
+        model = config.model.name
+
+    tokenizer = model.tokenizer if hasattr(model, 'tokenizer') else None
+
+    # The dataset to use for evaluation
+
+    if config.hf_model:
```
Comment (coryMosaicML) [inline diffusion/generate.py:40]: Is it possible to use hf datasets with non-hf models and vice versa? If so, using a `hf_model` flag to determine the dataset type might be confusing/limiting.

**Review by coryMosaicML — APPROVED** [review]
Looks good! Left a few minor comments/questions.

File: diffusion/evaluation/generate_images.py (hunk 2 above)
Comment (rishab-partha) [reply diffusion/evaluation/generate_images.py:65]: deleted

File: diffusion/evaluation/generate_images.py (hunk 3 above)
Comment (rishab-partha) [reply diffusion/evaluation/generate_images.py:66]: yup, also added to the dataset loading logic

File: diffusion/evaluation/generate_images.py (hunk 4 above)
Comment (rishab-partha) [reply diffusion/evaluation/generate_images.py:120]: yes, addressed in below comment

File: diffusion/generate.py (hunk 5 above)
Comment (rishab-partha) [reply diffusion/generate.py:40]: added a new flag `hf_dataset` to delineate the difference

File: diffusion/evaluation/generate_images.py (hunk 1 above)
Comment (rishab-partha) [reply diffusion/evaluation/generate_images.py:59]: added

**Pull request merged**